### PR TITLE
Preprocess fn

### PR DESCRIPTION
**Pull request by behrica**
This is a proposal to allow grid searching over preprocessing parameters, as discussed here:

https://clojurians.zulipchat.com/#narrow/stream/236259-tech.2Eml.2Edataset.2Edev/topic/introduce.20concept.20of.20preprocess.20fn.20into.20tech.2Eml

### CODE REVIEW

File: src/tech/v3/ml.clj [hunk 1]
```diff
@@ -230,6 +243,7 @@ see tech.v3.dataset.modelling/set-inference-target")
      (->> gs-seq
           (pmap #(do-k-fold % loss-fn target-colname ds-seq))
           (sort-by :avg-loss)
-          (take n-result-models))))
+          (take n-result-models)
```
Comment (behrica) [inline src/tech/v3/ml.clj:246]: This is a unrelated bug fix.
The originalò code did not restrict the number of models returned to `n-result-models'

File: test/tech/v3/ml_test.clj [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  (def dataset
+    (->
+     (ds/->dataset "test/data/reviews.csv.gz" {:key-fn keyword })
+     (ds/select-columns [:Text :Score])
+     (ds/update-column :Score #(map dec %))
+     (nlp/count-vectorize :Text :bow nlp/default-text->bow)
+     (ds-mod/set-inference-target :Score)))
+
+
+  (defn preprocess [ds options]
```
Comment (behrica) [inline test/tech/v3/ml_test.clj]: This is the pipeline specific preprocess function, adn it want to grid search over parameter "vocab-size"

File: test/tech/v3/ml_test.clj [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    (-> ds
+        (nb/bow->SparseArray :bow :bow-sparse (:vocab-size options))))
+
+  (def models
```
Comment (behrica) [inline test/tech/v3/ml_test.clj]: usage in grid search

File: test/tech/v3/ml_test.clj [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    (ml/train-auto-gridsearch dataset {:model-type :discrete-naive-bayes
+                                       :discrete-naive-bayes-model :multinomial
+                                       :sparse-column :bow-sparse
+                                       :k 5
+                                       :preprocess {:fn preprocess
+                                                    :vocab-size (ml-gs/linear 100 10000)}
+
+                                       }))
+
+  (ml/train dataset {:model-type :discrete-naive-bayes
```
Comment (behrica) [inline test/tech/v3/ml_test.clj]: usage in `train`, predict is teh same,

File: src/tech/v3/ml.clj [hunk 5]
```diff
@@ -49,6 +52,13 @@
   (:hyperparameters (options->model-def {:model-type model-kwd})))
 
 
+(defn preprocess [dataset options]
```
Comment (behrica) [inline src/tech/v3/ml.clj:60]: read and apply the preprocess function from the options

**Comment by behrica** [comment]
The last commit allows the preprocess-fn to contribute options.

File: src/tech/v3/ml.clj [hunk 6]
```diff
@@ -49,6 +52,12 @@
...
+  (let [fun (get-in options [:preprocess :fn] identity)]
```
Comment (behrica) [inline src/tech/v3/ml.clj]: Maybe the ":preprocess" submap does not make sense here, and the pre-processing related keys should be on top level of the options map ?

Comment (cnuernber) [reply src/tech/v3/ml.clj]: I would like them to be at the top level.  There aren't that many keys in the map as is.  I expect most people will want to serialize options with the model, so whatever is in the options map will need to be, in some cases serializable.

I think if :preprocess-fn can also be a namespaced symbol or keyword we can do 'requiring-resolve' to resolve to the actual preprocess-fn.  This also allows it to be used in a distributive fashion.

Comment (behrica) [reply src/tech/v3/ml.clj]: I did not think about distributed execution.
Is there some infrastructure or example for this somewhere ?

Comment (behrica) [reply src/tech/v3/ml.clj]: I have implemented it, as you suggested
